### DISCUSSION TITLE
differentiate %2F and / in path

### DISCUSSION THREAD

**Discussion by link89** [Kludex/uvicorn#2423] [Potential Issue] [answered]
According to this disccusion in [SO](https://stackoverflow.com/questions/1957115/is-a-slash-equivalent-to-an-encoded-slash-2f-in-the-path-portion-of-a),  `%2f` and `/` is different in the example of  [W3C URI recommendations](https://www.w3.org/Addressing/URL/4_URI_Recommentations.html)
> Example 1
> The URIs
> http://www.w3.org/albert/bertram/marie-claude
>
> and
> http://www.w3.org/albert/bertram/marie%2Dclaude
>
> are identical, as the %2D encodes a hyphen character.
>
> Example 2
> The URIs
> http://www.w3.org/albert/bertram/marie-claude
>
> and
> http://www.w3.org/albert/bertram%2Fmarie-claude
> 
> are NOT identical, as in the second case the encoded slash does not have hierarchical significance.

For short,  `%2F` should not be unquoted when unquoting path.

There are a lot of issue related to this issue  in different project:
* https://github.com/fastapi/fastapi/issues/791
* https://github.com/encode/starlette/issues/826
* https://github.com/fastapi/fastapi/issues/4390

Those problems need to be fixed in the asgi.

**Comment by link89** [comment]
I have created a PR to show what actually means.

https://github.com/encode/uvicorn/pull/2424

**Comment by link89** [comment]
Hi @Kludex let's move our discussion here. I have post some issues related to this problem. In my case switch to `hypecorn` can solve the issue, but since `uvicorn` is more popular, I think it is worth to fix it.

**Reply by Kludex** [reply]
Can you point me hypercorn's code source that has this logic?

**Reply by link89** [reply]
I guess I made an embarrassing mistake during testing. When I test `hypercorn` the pattern in FastAPI is set to `/doi/{doi:path}` instead of `/doi/{doi}`.

But I think that is still a valid point to leave `%2F` without unquote. 
When a user fire a request with a path parameter, 
if they write `"/doi/" + quote(doi)` where `doi = '10.14454/jcam.123'`, 
it is clear that they want the server to treat it as a parameter, 
or else they can join without quote.

**Comment by vvanglro** [comment, marked as answer]
It seems to be the same logic in the handling of Path.

https://github.com/pgjones/hypercorn/blob/main/src/hypercorn/protocol/http_stream.py#L93
https://github.com/pgjones/hypercorn/blob/main/src/hypercorn/protocol/ws_stream.py#L215

**Reply by link89** [reply]
Oops, I think I made some mistake
